File: skill.js
#!/usr/bin/env node
/**
 * Google News RSS Feed Skill
 * Author: Valerius Mattfeld (valerius21)
 * Usage: skill.js <query> [count] [region]
 * Example: skill.js "legal tech" 10 US
 */

const https = require("https");

function httpsGet(url, options) {
  return new Promise((resolve, reject) => {
    https
      .get(url, options, (res) => {
        if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          httpsGet(res.headers.location, options).then(resolve, reject);
          return;
        }
        let data = "";
        res.on("data", (chunk) => (data += chunk));
        res.on("end", () => resolve(data));
      })
      .on("error", reject);
  });
}

async function fetchGoogleNews(query, count = 5, region = "US") {
  if (!query) {
    console.log("Usage: google-news <query> [count] [region]");
    console.log('Example: google-news "legal tech" 10 US');
    console.log();
    console.log("Parameters:");
    console.log("  query   - Search term (required)");
    console.log("  count   - Number of results (default: 5)");
    console.log("  region  - Country code (default: US, e.g., DE, GB, FR)");
    process.exit(1);
  }

  const encodedQuery = encodeURIComponent(query);
  const rssUrl = `https://news.google.com/rss/search?hl=en&gl=${region}&ceid=${region}%3Aen&q=${encodedQuery}`;

  console.log(`Google News: "${query}" (Region: ${region})`);
  console.log(rssUrl);
  console.log();

  const options = {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
    },
  };

  try {
    const data = await httpsGet(rssUrl, options);
    const titles = [];
    const itemRegex = /<item>[\s\S]*?<\/item>/g;
    const titleRegex = /<title><!\[CDATA\[(.*?)\]\]>|<title>(.*?)<\/title>/;

    let match;
    while ((match = itemRegex.exec(data)) !== null) {
      const titleMatch = match[0].match(titleRegex);
      if (titleMatch) {
        const title = (titleMatch[1] || titleMatch[2] || "").trim();
        if (
          title &&
          title.toLowerCase() !== query.toLowerCase() &&
          title.toLowerCase() !== "google news"
        ) {
          titles.push(title);
        }
      }
    }

    for (let i = 0; i < Math.min(titles.length, count); i++) {
      console.log(` ${String(i + 1).padStart(2)}. ${titles[i]}`);
    }

    if (titles.length === 0) {
      console.log(" No results found.");
    }

    console.log();
    console.log(`Tip: Add this URL to your blogwatcher: ${rssUrl}`);
  } catch (err) {
    console.error(` Error fetching RSS feed: ${err.message}`);
    process.exit(1);
  }
}

const query = process.argv[2] || "";
const count = parseInt(process.argv[3], 10) || 5;
const region = process.argv[4] || "US";

fetchGoogleNews(query, count, region);

File: skill.py
#!/usr/bin/env python3
"""
Google News RSS Feed Skill
Author: Valerius Mattfeld (valerius21)
Usage: skill.py <query> [count] [region]
Example: skill.py "legal tech" 10 US
"""

import sys
import urllib.parse
import urllib.request
import xml.etree.ElementTree as ET


def fetch_google_news(query: str, count: int = 5, region: str = "US") -> None:
    """Fetch and display Google News RSS feed for a given query."""
    
    if not query:
        print("Usage: google-news <query> [count] [region]")
        print("Example: google-news \"legal tech\" 10 US")
        print()
        print("Parameters:")
        print("  query   - Search term (required)")
        print("  count   - Number of results (default: 5)")
        print("  region  - Country code (default: US, e.g., DE, GB, FR)")
        sys.exit(1)
    
    encoded_query = urllib.parse.quote(query, safe='')
    rss_url = f"https://news.google.com/rss/search?hl=en&gl={region}&ceid={region}%3Aen&q={encoded_query}"
    
    print(f"Google News: \"{query}\" (Region: {region})")
    print(f"{rss_url}")
    print()
    
    try:
        req = urllib.request.Request(
            rss_url,
            headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'}
        )
        
        with urllib.request.urlopen(req, timeout=10) as response:
            xml_content = response.read().decode('utf-8')
        
        root = ET.fromstring(xml_content)
        items = root.findall('.//item')
        titles = []
        
        for item in items:
            title_elem = item.find('title')
            if title_elem is not None and title_elem.text:
                title = title_elem.text.strip()
                if title and title.lower() != query.lower() and title.lower() != "google news":
                    titles.append(title)
        
        for i, title in enumerate(titles[:count], 1):
            print(f" {i:2d}. {title}")
        
        if not titles:
            print(" No results found.")
        
    except urllib.error.URLError as e:
        print(f" Error fetching RSS feed: {e}")
        sys.exit(1)
    except ET.ParseError as e:
        print(f" Error parsing XML: {e}")
        sys.exit(1)
    except Exception as e:
        print(f" Unexpected error: {e}")
        sys.exit(1)
    
    print()
    print(f"Tip: Add this URL to your blogwatcher: {rss_url}")


def main():
    """Main entry point."""
    query = sys.argv[1] if len(sys.argv) > 1 else ""
    count = int(sys.argv[2]) if len(sys.argv) > 2 else 5
    region = sys.argv[3] if len(sys.argv) > 3 else "US"
    
    fetch_google_news(query, count, region)


if __name__ == "__main__":
    main()
